python/trainers/buffer.py

```diff
@@ -0,0 +1,233 @@
+			:param key_list: The fields which length will be compared
+			"""
+			if len(key_list) < 2:
+				return True
+			l = None
+			for key in key_list:
+				if key not in self.keys():
+					return False
+				if ((l != None) and (l!=len(self[key]))):
+					return False
+				l = len(self[key])
+			return True
+		def shuffle(self, key_list = None):
+			"""
+			Shuffles the fields in key_list in a consistent way: The reordering will
+			be the same accross fields.
+			:param key_list: The fields that must be shuffled.
+			"""
+			if key_list is None:
+				key_list = list(self.keys())
+			if not self.check_length(key_list):
+				raise BufferException("Unable to shuffle if the fields are not of same length")
+				return
+			s = np.arange(len(self[key_list[0]]))
+			np.random.shuffle(s)
+			for key in key_list:
+				self[key][:] = [self[key][i] for i in s]
+				#memory leak here ?
+				# with l as [self[key][i] for i in s]:
+				# 	self[key].reset_field()
+				# 	self[key] += l
+				# self[key] = Buffer.AgentBuffer.AgentBufferField([self[key][i] for i in s])
+				# self[key].reorder(s)
+	def __init__(self):
+		self.global_buffer = self.AgentBuffer() 
+		super(Buffer, self).__init__()
+	def __str__(self):
+		return "global buffer :\n\t{0}\nlocal_buffers :\n{1}".format(str(self.global_buffer), 
+			'\n'.join(['\tagent {0} :{1}'.format(k, str(self[k])) for k in self.keys()]))
+	def __getitem__(self, key):
+		if key not in self.keys():
+			self[key] = self.AgentBuffer()
+		return super(Buffer, self).__getitem__(key)
+	def append_BrainInfo(self, info):
+		"""
+		Appends the information in a BrainInfo element to the buffer.
+		:param info: The BrainInfo from which the information will be extracted.
+		"""
+		raise BufferException("This method is not yet implemented")
+		# TODO: Find how useful this would be
+		# TODO: Implementation
+	def reset_global(self):
+		"""
+		Resets the global buffer
+		"""
+		self.global_buffer.reset_agent()
+	def reset_all(self):
+		"""
+		Resets the global buffer and all the local local_buffers
+		"""
+		# TODO: Need a better name
+		self.global_buffer.reset_agent()
+		agent_ids = list(self.keys())
+		for k in agent_ids:
+			self[k].reset_agent()
+	def append_global(self, agent_id ,key_list = None,  batch_size = None, training_length = None):
+		"""
+		Appends the buffer of an agent to the global buffer.
+		:param agent_id: The id of the agent which data will be appended
+		:param key_list: The fields that must be added. If None: all fields will be appended.
+		:param batch_size: The number of elements that must be appended. If None: All of them will be.
+		:param training_length: The length of the samples that must be appended. If None: only takes one element.
+		"""
+		if key_list is None:
+			key_list = self[agent_id].keys()
+		if not self[agent_id].check_length(key_list):
+			raise BufferException("The length of the fields {0} for agent {1} where not of comparable length"
+				.format(key_list, agent_id))
```

Comment:
consider using f-strings in python 3.6+

```python
raise BufferException(f'The length of the fields {key_list} for agent {agent_id} where not of comparable length')
```

https://www.python.org/dev/peps/pep-0498/#abstract

p.s. that message is ambiguous because lengths can always be compared. what does it mean for those to not be comparable as the exception claims?